async-nats/src/error/mod.rs

```diff
@@ -0,0 +1,160 @@
+use std::fmt::{Debug, Display};
+
+/// The error type for the NATS client, generic by the kind of error.
+#[derive(Debug)]
+pub struct Error<Kind>
+where
+    Kind: Clone + Debug + Display + PartialEq,
+{
+    pub(crate) kind: Kind,
+    pub(crate) source: Option<crate::Error>,
+}
+
+impl<Kind> Error<Kind>
+where
+    Kind: Clone + Debug + Display + PartialEq,
+{
+    pub(crate) fn new(kind: Kind) -> Self {
+        Self { kind, source: None }
+    }
+
+    pub(crate) fn with_source<S>(kind: Kind, source: S) -> Self
+    where
+        S: Into<crate::Error>,
+    {
+        Self {
+            kind,
+            source: Some(source.into()),
+        }
+    }
+
+    // In some cases the kind doesn't implement `Copy` trait
+    pub fn kind(&self) -> Kind {
```

Comment:
This can just return ref? no clone needed.

```suggestion
    pub fn kind(&self) -> &Kind {
```

Reply:
I asked myself the same question, but left this method for the backward compatibitlity only. Without it we could just make the `kind` attribute public ))

Reply:
Its Copy most of the time. Clone only in case or JetStream error. And if you look at std::io::Error its also returning copy.

Reply:
Not only the `JetStream`, but also a Request keeping a string inside, which is not copy.